zerver/tests/test_auth_backends.py

```diff
@@ -2828,6 +2829,125 @@ def test_social_auth_no_key(self) -> None:
         pass
 
 
+class GenericOpenIdConnectTest(SocialAuthBase):
+    __unittest_skip__ = False
+
+    BACKEND_CLASS = GenericOpenIdConnectBackend
+    CLIENT_KEY_SETTING = "SOCIAL_AUTH_TESTOIDC_KEY"
+    CLIENT_SECRET_SETTING = "SOCIAL_AUTH_TESTOIDC_SECRET"
+    LOGIN_URL = "/accounts/login/social/testoidc"
+    SIGNUP_URL = "/accounts/register/social/testoidc"
+
+    BASE_OIDC_URL = "https://example.com/api/openid"
+    AUTHORIZATION_URL = f"{BASE_OIDC_URL}/authorize"
+    ACCESS_TOKEN_URL = f"{BASE_OIDC_URL}/token"
+    JWKS_URL = f"{BASE_OIDC_URL}/jwks"
+    USER_INFO_URL = f"{BASE_OIDC_URL}/userinfo"
+    AUTH_FINISH_URL = "/complete/testoidc/"
+    CONFIG_ERROR_URL = "/config-error/testoidc"
+
+    def social_auth_test(
+        self,
+        *args: Any,
+        **kwargs: Any,
+    ) -> HttpResponse:
+        idp_discovery_endpoint_payload_dict = {
+            "issuer": self.BASE_OIDC_URL,
+            "authorization_endpoint": self.AUTHORIZATION_URL,
+            "token_endpoint": self.ACCESS_TOKEN_URL,
+            "userinfo_endpoint": self.USER_INFO_URL,
+            "response_types_supported": [
+                "code",
+                "id_token",
+                "id_token token",
+                "code token",
+                "code id_token",
+                "code id_token token",
+            ],
+            "jwks_uri": self.JWKS_URL,
+            "id_token_signing_alg_values_supported": ["HS256", "RS256"],
+            "subject_types_supported": ["public"],
+            "token_endpoint_auth_methods_supported": ["client_secret_post", "client_secret_basic"],
+        }
+
+        # We need to run the social_auth_test procedure with a mock response set up for the
+        # OIDC discovery endpoint as that's the first thing requested by the server when a user
+        # starts trying to authenticate.
+        with responses.RequestsMock(assert_all_requests_are_fired=False) as requests_mock:
+            requests_mock.add(
+                requests_mock.GET,
+                f"{self.BASE_OIDC_URL}/.well-known/openid-configuration",
+                match_querystring=False,
+                status=200,
+                body=json.dumps(idp_discovery_endpoint_payload_dict),
+            )
+            result = super().social_auth_test(*args, **kwargs)
+
+        return result
+
+    def social_auth_test_finish(self, *args: Any, **kwargs: Any) -> HttpResponse:
+        # Trying to generate a (access_token, id_token) pair here in tests that would
+        # successfully pass validation by validate_and_return_id_token is impractical
+        # and unnecessaru (see upstream implementation of the method for how the validation works).
```

Comment:
Typo in "unecessary".  For "upstream", do you mean in python-social-auth?  Probably worth specifying just for clarity.

Reply:
yeah, fixed